### PR TITLE
NXP-23291: Add audit storage interface

### PR DESCRIPTION
**Pull request by funshodavid**

### CODE REVIEW

**Comment by funshodavid** [comment]
T&P running: https://qa.nuxeo.org/jenkins/view/TestAndPush/job/TestAndPush/job/ondemand-testandpush-bdelbosc2/150

**Comment by bdelbosc** [comment]
:+1:

File: nuxeo-features/nuxeo-elasticsearch/nuxeo-elasticsearch-audit/src/test/java/org/nuxeo/elasticsearch/TestAuditWithElasticSearch.java [hunk 1]
```diff
@@ -244,5 +253,44 @@ public void testGetLogEntriesAfter() throws Exception {
...
+        LogEntryGen.flushAndSync();
+
+        // Query all logs
+        AuditQueryBuilder builder = new AuditQueryBuilder().predicates(Predicates.eq("eventId", "idForAuditStorage"));
+        // builder.predicates()
+        List<LogEntry> logs = esBackend.queryLogs(builder);
+        assertEquals(42, logs.size());
+
+        ScrollResult scrollResult = esBackend.scroll(builder, 5, 10);
+        assertEquals(5, scrollResult.getResultIds().size());
```
Comment (ataillefer) [inline nuxeo-features/nuxeo-elasticsearch/nuxeo-elasticsearch-audit/src/test/java/org/nuxeo/elasticsearch/TestAuditWithElasticSearch.java]: Could be inside the `while` no?

File: nuxeo-features/nuxeo-platform-audit/nuxeo-platform-audit-core/src/test/java/org/nuxeo/ecm/platform/audit/TestAuditStorage.java [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+        backend.append(jsonEntries);
+
+        // Query all logs
+        AuditQueryBuilder builder = new AuditQueryBuilder().predicates(Predicates.eq("eventId", "idForAuditStorage"));
+        // builder.predicates()
+        List<LogEntry> logs = backend.queryLogs(builder);
+        assertEquals(42, logs.size());
+
+        ScrollResult scrollResult = backend.scroll(builder, 5, 10);
+        assertEquals(5, scrollResult.getResultIds().size());
```
Comment (ataillefer) [inline nuxeo-features/nuxeo-platform-audit/nuxeo-platform-audit-core/src/test/java/org/nuxeo/ecm/platform/audit/TestAuditStorage.java]: Same remark

File: nuxeo-features/nuxeo-elasticsearch/nuxeo-elasticsearch-audit/src/test/java/org/nuxeo/elasticsearch/TestAuditWithElasticSearch.java (hunk 1 above)
Comment (funshodavid) [reply nuxeo-features/nuxeo-elasticsearch/nuxeo-elasticsearch-audit/src/test/java/org/nuxeo/elasticsearch/TestAuditWithElasticSearch.java]: The last scroll results could contain less than 5 entries (in this test for example the last retrieved results will contain 2 entries).

Comment (ataillefer) [reply nuxeo-features/nuxeo-elasticsearch/nuxeo-elasticsearch-audit/src/test/java/org/nuxeo/elasticsearch/TestAuditWithElasticSearch.java]: Right, so could check count <= 5, or not :smile:

File: nuxeo-features/nuxeo-platform-audit/nuxeo-platform-audit-core/src/main/java/org/nuxeo/ecm/platform/audit/service/AbstractAuditBackend.java [hunk 3]
```diff
@@ -460,4 +470,31 @@ protected Predicate convert(FilterMapEntry entry) {
...
+        }).collect(Collectors.toList());
+        addLogEntries(logEntries);
+    }
+
+    @Override
+    public ScrollResult scroll(AuditQueryBuilder queryBuilder, int batchSize, int keepAlive) {
+
+        cursorService.checkForTimedOutScroll();
+        List<LogEntry> logEntries = queryLogs(queryBuilder);
+        String scrollId = cursorService.registerCursor(logEntries.iterator(), batchSize, keepAlive);
```
Comment (kevinleturc) [inline nuxeo-features/nuxeo-platform-audit/nuxeo-platform-audit-core/src/main/java/org/nuxeo/ecm/platform/audit/service/AbstractAuditBackend.java]: Here you're gonna load probably a lot of entries in memory, the scroll will only be on the local `List`. Does it what we want ?

Comment (funshodavid) [reply nuxeo-features/nuxeo-platform-audit/nuxeo-platform-audit-core/src/main/java/org/nuxeo/ecm/platform/audit/service/AbstractAuditBackend.java]: We can use it as a default implementation, but the next step is to have each backend overriding it to use its own scrolling API if possible (e.g Elasticsearch).

Comment (kevinleturc) [reply nuxeo-features/nuxeo-platform-audit/nuxeo-platform-audit-core/src/main/java/org/nuxeo/ecm/platform/audit/service/AbstractAuditBackend.java]: 👍

**Comment by kevinleturc** [comment]
👍

**Pull request merged**